File: servidor.js
#!/usr/bin/env node
/*
 * Servidor local do site Phygital Brasil — versão Node.
 *
 * Mesmo comportamento do servidor.py: envia Cache-Control: no-store em tudo,
 * porque sem isso o navegador guarda o CSS e o JavaScript antigos e as
 * alterações não aparecem sem limpar o cache na mão.
 *
 * Existe porque o Windows não traz Python instalado. Em máquinas com Python,
 * servidor.py faz exatamente a mesma coisa.
 *
 * Uso:  node servidor.js [porta]
 */

const http = require('http');
const fs = require('fs');
const path = require('path');

const RAIZ = path.join(__dirname, 'site');

const TIPOS = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.webp': 'image/webp',
  '.ico': 'image/x-icon',
  '.woff2': 'font/woff2',
  '.xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  '.zip': 'application/zip'
};

function servir(req, res) {
  let caminho = decodeURIComponent(req.url.split('?')[0].split('#')[0]);
  if (caminho.endsWith('/')) caminho += 'index.html';

  const alvo = path.join(RAIZ, path.normalize(caminho));

  /* Impede sair da pasta site/ por ../ na URL */
  if (!alvo.startsWith(RAIZ)) {
    res.writeHead(403).end('403');
    return;
  }

  fs.readFile(alvo, (erro, dados) => {
    if (erro) {
      res.writeHead(404, { 'Content-Type': 'text/html; charset=utf-8' });
      res.end('<h1>404</h1><p>' + caminho + '</p>');
      console.log('404', caminho);
      return;
    }
    res.writeHead(200, {
      'Content-Type': TIPOS[path.extname(alvo).toLowerCase()] || 'application/octet-stream',
      'Cache-Control': 'no-store, no-cache, must-revalidate, max-age=0',
      'Pragma': 'no-cache',
      'Expires': '0'
    });
    res.end(dados);
  });
}

function subir(porta, tentativasRestantes) {
  const servidor = http.createServer(servir);

  servidor.on('error', (e) => {
    if (e.code === 'EADDRINUSE' && tentativasRestantes > 0) {
      subir(porta + 1, tentativasRestantes - 1);
    } else {
      console.error('Não foi possível subir o servidor:', e.message);
      process.exit(1);
    }
  });

  servidor.listen(porta, () => {
    const url = 'http://localhost:' + porta;
    console.log('');
    console.log('  ██  PHYGITAL BRASIL — SERVIDOR LOCAL');
    console.log('  ' + '─'.repeat(48));
    console.log('  Site:              ' + url);
    console.log('  Painel Competidor: ' + url + '/painel/');
    console.log('  Painel Admin:      ' + url + '/admin/');
    console.log('  ' + '─'.repeat(48));
    console.log('  Cache desativado: basta recarregar a página (Ctrl+R)');
    console.log('  Para parar: Ctrl+C');
    console.log('');
  });
}

if (!fs.existsSync(RAIZ)) {
  console.error("Pasta 'site' não encontrada em " + RAIZ);
  process.exit(1);
}

const pedida = parseInt(process.argv[2], 10);
subir(Number.isInteger(pedida) ? pedida : 8080, 12);

File: servidor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Servidor local do site Phygital Brasil.

Igual ao `python3 -m http.server`, com uma diferença importante: envia
Cache-Control: no-store em tudo. Sem isso o navegador guarda o CSS e o
JavaScript antigos e as alterações não aparecem sem limpar o cache na mão.

Uso:  python3 servidor.py [porta]
"""

import functools
import http.server
import os
import socketserver
import sys
import webbrowser

RAIZ = os.path.join(os.path.dirname(os.path.abspath(__file__)), "site")


class SemCache(http.server.SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header("Cache-Control", "no-store, no-cache, must-revalidate, max-age=0")
        self.send_header("Pragma", "no-cache")
        self.send_header("Expires", "0")
        super().end_headers()

    def log_message(self, formato, *args):
        # Só registra erros: o log de cada arquivo servido polui o terminal.
        codigo = str(args[1]) if len(args) > 1 else ""
        if codigo.startswith(("4", "5")):
            super().log_message(formato, *args)


def porta_livre(inicial=8080, tentativas=12):
    for p in range(inicial, inicial + tentativas):
        try:
            with socketserver.TCPServer(("", p), None):
                return p
        except OSError:
            continue
    return None


def main():
    if not os.path.isdir(RAIZ):
        print("Pasta 'site' não encontrada em", RAIZ)
        return 1

    pedida = int(sys.argv[1]) if len(sys.argv) > 1 and sys.argv[1].isdigit() else 8080
    porta = porta_livre(pedida)
    if porta is None:
        print("Nenhuma porta livre a partir de", pedida)
        return 1

    url = "http://localhost:%d" % porta
    handler = functools.partial(SemCache, directory=RAIZ)

    print("")
    print("  ██  PHYGITAL BRASIL — SERVIDOR LOCAL")
    print("  " + "─" * 48)
    print("  Site:              %s" % url)
    print("  Painel Competidor: %s/painel/" % url)
    print("  Painel Admin:      %s/admin/" % url)
    print("  " + "─" * 48)
    print("  Cache desativado: basta recarregar a página (Cmd+R)")
    print("  Para parar: Ctrl+C")
    print("")

    socketserver.TCPServer.allow_reuse_address = True
    with socketserver.TCPServer(("", porta), handler) as servidor:
        try:
            webbrowser.open(url)
        except Exception:
            pass
        try:
            servidor.serve_forever()
        except KeyboardInterrupt:
            print("\nServidor encerrado.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
